pkg/controller/machine/machine.go

```diff
@@ -585,50 +527,11 @@ func (c *Controller) deleteNodeForMachine(machine *clusterv1alpha1.Machine) erro
 
 func (c *Controller) ensureInstanceExistsForMachine(prov cloud.Provider, machine *clusterv1alpha1.Machine, userdataProvider userdata.Provider, providerConfig *providerconfig.Config) error {
 	glog.V(6).Infof("Requesting instance for machine '%s' from cloudprovider because no associated node with status ready found...", machine.Name)
-	// case 1: validate the machine spec before getting the instance from cloud provider.
-	// even though this is a little bit premature and inefficient, it helps us detect invalid specification
-	defaultedMachineSpec, changed, err := prov.AddDefaults(machine.Spec)
-	if err != nil {
-		return c.updateMachineErrorIfTerminalError(machine, common.InvalidConfigurationMachineError, err.Error(), err, "failed to add defaults to machine")
-	}
-	if changed {
-		glog.V(4).Infof("updating machine '%s' with defaults...", machine.Name)
-		c.recorder.Event(machine, corev1.EventTypeNormal, "Defaulted", "Updated machine with defaults")
-		if machine, err = c.updateMachine(machine, func(m *clusterv1alpha1.Machine) {
-			m.Spec = defaultedMachineSpec
-		}); err != nil {
-			return fmt.Errorf("failed to update machine '%s' after adding defaults: '%v'", machine.Name, err)
-		}
 
-		glog.V(4).Infof("Successfully updated machine '%s' with defaults!", machine.Name)
-	}
-
-	cacheKey := string(machine.UID) + machine.ResourceVersion
-	c.validationCacheMutex.Lock()
-	validationSuccess := c.validationCache[cacheKey]
-	c.validationCacheMutex.Unlock()
-	if !validationSuccess {
-		if err := c.validateMachine(prov, machine); err != nil {
-			if _, errNested := c.updateMachineError(machine, common.InvalidConfigurationMachineError, err.Error()); errNested != nil {
-				return fmt.Errorf("failed to update machine error after failed validation: %v", errNested)
-			}
-			return fmt.Errorf("invalid provider config: %v", err)
-		}
-		c.validationCacheMutex.Lock()
-		c.validationCache[cacheKey] = true
-		c.validationCacheMutex.Unlock()
-	} else {
-		glog.V(6).Infof("Skipping validation as the machine was already successfully validated before")
-	}
 	providerInstance, err := prov.Get(machine)
 
 	// case 2: retrieving instance from provider was not successful
 	if err != nil {
-		//First invalidate the validation cache to make sure we run the validation on the next sync.
-		//This might happen in case the user invalidates his provider credentials...
```

Comment:
Hm. The webhook can obviously not help if someone invalidates their credentials. Maybe we should keep the validation in the controller and just remove the defaulting?